rust_src/src/crypto/mod.rs

```diff
@@ -1,78 +1,470 @@
+        ];
+        let val = LispObject::from_raw(unsafe {
+            Ffind_operation_coding_system(4, args.as_mut_ptr())
+        });
+        if val.is_cons() && val.as_cons_or_error().cdr().is_not_nil() {
+            return val.as_cons_or_error().cdr();
+        }
+    }
+    if LispObject::from_raw(buffer.buffer_file_coding_system).is_not_nil() {
+        /* If we still have not decided a coding system, use the
+           default value of buffer-file-coding-system. */
+        return LispObject::from_raw(buffer.buffer_file_coding_system);
+    }
+    if fboundp(LispObject::from_raw(
+        unsafe { globals.f_Vselect_safe_coding_system_function },
+    )).is_not_nil()
+    {
+        /* Confirm that VAL can surely encode the current region. */
+        return LispObject::from_raw(unsafe {
+            call4(
+                globals.f_Vselect_safe_coding_system_function,
+                LispObject::from_natnum(start_byte as EmacsInt).to_raw(),
+                LispObject::from_natnum(end_byte as EmacsInt).to_raw(),
+                coding_system.to_raw(),
+                LispObject::constant_nil().to_raw(),
+            )
+        });
+    }
+    LispObject::constant_nil()
+}
+
+fn get_input_from_string(
+    object: LispObject,
+    string: LispStringRef,
+    start: LispObject,
+    end: LispObject,
+) -> LispObject {
+    let size: ptrdiff_t;
+    let start_byte: ptrdiff_t;
+    let end_byte: ptrdiff_t;
+    let mut start_char: ptrdiff_t = 0;
+    let mut end_char: ptrdiff_t = 0;
+
+    size = string.len_bytes();
+    unsafe {
+        validate_subarray(
+            object.to_raw(),
+            start.to_raw(),
+            end.to_raw(),
+            size,
+            &mut start_char,
+            &mut end_char,
+        );
+    }
+    start_byte = if start_char == 0 {
+        0
+    } else {
+        unsafe { string_char_to_byte(object.to_raw(), start_char) }
+    };
+    end_byte = if end_char == size {
+        string.len_bytes()
+    } else {
+        unsafe { string_char_to_byte(object.to_raw(), end_char) }
+    };
+    if start_byte == 0 && end_byte == size {
+        object
+    } else {
+        LispObject::from_raw(unsafe {
+            make_specified_string(
+                string.const_sdata_ptr().offset(start_byte),
+                -1 as ptrdiff_t,
+                end_byte - start_byte,
+                string.is_multibyte(),
+            )
+        })
+    }
+}
+
+fn get_input_from_buffer(
+    buffer: LispBufferRef,
+    start: LispObject,
+    end: LispObject,
+    start_byte: &mut ptrdiff_t,
+    end_byte: &mut ptrdiff_t,
+) -> LispObject {
+    let prev_buffer = unsafe { (*current_thread).m_current_buffer };
+    unsafe { record_unwind_current_buffer() };
+    unsafe { set_buffer_internal(buffer.as_ptr() as *const _ as *const libc::c_void) };
+    *start_byte = if start.is_nil() {
+        buffer.begv
+    } else {
+        match start.as_number_coerce_marker_or_error() {
```

Comment:
Hm, actually `as_number_coerce_marker_or_error` should not allow floats. I'll have to fix that.

That's a bit of inconsistent naming on the C side ("number" used instead of "fixnum" or "integer").